src/Request/RequestHandler.php

```diff
@@ -98,41 +95,117 @@ public function __construct(array $config)
+     * @param String $verb
+     * @param $method
+     * @param array $parameters
+     * @return array|string
+     * @throws ApiException
+     * @throws ValidationException
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    public function custom(String $verb, $method, $parameters = [])
+    {
+        // Only GET and POST are supported by PayFast
+        if ($verb !== 'GET' && $verb !== 'POST') {
+            throw new ValidationException(12100, sprintf('Invalid verb: %s', $verb));
+        }
+
+        if (is_null($method)) {
+            throw new ValidationException(12100, sprintf('Method cannot be null', $verb));
+        }
+
+        return $this->handleRequest($verb, $method, $parameters);
+    }
+
     /**
      * Makes a request using Guzzle
      *
-     * @param array $options Request options
+     * @param $verb
+     * @param $method
+     * @param array $parameters
      * @return array|string
      * @throws ApiException
      * @throws \GuzzleHttp\Exception\GuzzleException
      * @see RequestHandler::request()
      */
-    public function handleRequest(array $options)
+    private function handleRequest($verb, $method, array $parameters)
     {
+        // Get the timestamp in the required format
+        $date = new DateTime();
+        $date->setTimeZone(new DateTimeZone('UTC'));
+        $date->setTimestamp(time());
+        $timestamp = $date->format('Y-m-d\TH:i:s');
+
+        $signParameters = $this->getSigningObject($timestamp);
+
+        // Add the provided params to $signParameters
+        foreach ($parameters as $key => $value) {
+            $signParameters[$key] = $value;
+        }
+
+        // MD5 hash the fields
+        $signature = $this->generateSignature($signParameters);
+
+        // Create the headers
+        $request = [
+            'headers' => [
+                'Content-Type' => 'application/json',
+                'Content-Length' => 0,
+                'signature' => $signature,
+                'timestamp' => $timestamp,
+                'version' => self::VERSION,
+                'merchant-id' => $this->merchantId,
+            ],
+        ];
+
+        // Create the URI
         $protocol = $this->ssl ? 'https' : 'http';
 
         $uri = sprintf(
             '%s:%s:%s%s',
             $protocol,
             $this->endpoint,
             $this->port,
-            $this->method
+            $method
         );
 
+        // Are we using the sandbox?
         if ($this->testing === true) {
             $uri .= '?testing=true';
         }
 
+        if ($verb === 'POST') {
+            $data = json_encode($parameters);
+            $request['headers']['Content-Length'] = strlen($data);
+            $request['body'] = $data;
+        }
+
+        // We need to add params to URI
+        if ($verb === 'GET') {
+            if (!$this->testing) {
+                $uri .= '?';
+            } elseif (sizeof($parameters) > 0) {
+                $uri .= '&';
+            }
```

Comment:
These 5 lines seem a bit strange to me.

What's the aim here, and what is this code doing?

Reply:
So if it is testing and there are parameters, then the uri will look something like `url/history?testing=true` and then we add a `&`, and then add the params, if it isn't testing, we first need to add the `?`, then the params
I agree this isn't as elegant as it could be, any tips?